src/main/java/org/broadinstitute/hellbender/metrics/MultiLevelCollectorReducible.java

```diff
@@ -0,0 +1,428 @@
+
+    // Convert the current SAMRecord and the ReferenceSequence for that record into an ARGTYPE object
+    // see accept record for use
+    protected abstract ARGTYPE makeArg(final SAMRecord samRec, final ReferenceSequence refSeq);
+
+    /**
+     * Construct a PerUnitMetricCollector with the given arguments.
+     * @param sample If aggregating by ALL_READS this will be null, otherwise the sample that will be used to identify
+     *               this collector
+     * @param library If aggregating by SAMPLE this will be null, otherwise the library that will be used to identify
+     *               this collector
+     * @param readGroup If aggregating by LIBRARY this will be null, otherwise the readGroup that will be used to identify
+     *                  this collector
+     * @return A PerUnitMetricCollector parameterized by the given arguments
+     */
+    protected abstract UNIT_COLLECTOR makeChildCollector(final String sample, final String library, final String readGroup);
+
+    //These are exposed here (rather than being encapsulated in the Distributor subclasses below in order
+    //to provide subclasses with an explicit point to add initialization (specific to accumulation level) for
+    //a PerUnitMetricCollector it is creating
+    protected UNIT_COLLECTOR makeAllReadCollector() {
+        return makeChildCollector(null, null, null);
+    }
+    protected UNIT_COLLECTOR makeSampleCollector(final SAMReadGroupRecord rg) {
+        return makeChildCollector(rg.getSample(), null, null);
+    }
+    protected UNIT_COLLECTOR makeLibraryCollector(final SAMReadGroupRecord rg) {
+        return makeChildCollector(rg.getSample(), rg.getLibrary(), null);
+    }
+    protected UNIT_COLLECTOR makeReadGroupCollector(final SAMReadGroupRecord rg) {
+        return makeChildCollector(rg.getSample(), rg.getLibrary(), rg.getPlatformUnit());
+    }
+
+    /**
+     * Combine/merge a source MultiLevelCollectorReducible object into this MultiLevelCollectorReducible.
+     *
+     * Most derived classes will want to provide a type-specific version of this method that delegates
+     * to this method, but is typed suitably for use as a JavaRDD.reduce combiner method. Given a
+     * derived class of type class "D", the combine method looks like this:
+     *
+     * public D combine(D target, D source) {
+     *     target.combine(source);  // calls this implementation
+     *     return target;           // return an object of type D
+     * }
+     *
+     * Such a wrapper can be passed directly to JavaRDD.reduce.
+     *
+     * @param source source MultiLevelCollectorReducible to be combined with this object
+     */
+    public void combine(MultiLevelCollectorReducible<METRIC_TYPE, HISTOGRAM_KEY, ARGTYPE, UNIT_COLLECTOR> source)
+    {
+        if ((allReadCollector == null) != (source.allReadCollector == null)) {
+            throw new IllegalArgumentException("MultiLevelCollectors must be the same size to be combined");
```

Comment:
somehow this messase seems weird given that it checks nullness of allReadCollector